pennylane/optimize/rotosolve.py

```diff
@@ -12,172 +12,411 @@
-        x_new = self.step(objective_fn, *args, **kwargs)
-
-        return x_new, objective_fn(*args, **kwargs)
-
-    def step(self, objective_fn, *args, **kwargs):
-        r"""Update args with one step of the optimizer.
-
         Args:
             objective_fn (function): the objective function for optimization. It should take a
-                sequence of the values ``*args`` and a list of the gates ``generators`` as inputs, and
-                return a single value.
+                sequence of the values ``*args`` and a list of the gates ``generators`` as inputs,
+                and return a single value.
             *args : variable length sequence containing the initial
                 values of the variables to be optimized over or a single float with the initial
                 value.
+            num_frequencies (int|array[int]): The number of frequencies in the ``objective_fn`` per
+                parameter. If an ``int``, the same number is used for all parameters; if
+                ``array[int]``, the shape of ``args`` and ``num_frequencies`` has to coincide.
+                Defaults to ``num_frequencies=1``, corresponding to Pauli rotation gates.
+            optimizer (callable|"brute"|"shgo"): the optimization method used for the univariate
+                minimization. If a callable, should have the signature
+                ``(fun, **kwargs) -> x_min, y_min``, where ``y_min`` is tracked and returned
+                if ``full_output==True`` but is not relevant to the optimization.
+                If ``"brute"`` or ``"shgo"``, the corresponding global optimizer of SciPy is used.
+                Defaults to ``"brute"``.
+            optimizer_kwargs : keyword arguments for the ``optimizer``. For ``"brute"`` and
+                ``"shgo"``, these kwargs are passed to the respective SciPy implementation.
+                Has to be given as one dictionary, *not variable length*.
+            full_output (bool): whether to return the intermediate minimized energy values from
+                the univariate optimization steps.
             **kwargs : variable length keyword arguments for the objective function.
 
         Returns:
             list [array]: the new variable values :math:`x^{(t+1)}`.
             If single arg is provided, list [array] is replaced by array.
         """
-        # will single out one variable to change at a time
+        if num_frequencies is None:
+            num_frequencies = [1] * len(args)
+        elif np.isscalar(num_frequencies) and np.isclose(int(num_frequencies), num_frequencies):
+            num_frequencies = [num_frequencies] * len(args)
+        else:
+            if len(num_frequencies) != len(args):
+                raise ValueError(
+                    "The number of the provided numbers of frequencies "
+                    f"({len(num_frequencies)}) does not match the number of function arguments "
+                    f"({len(args)})."
+                )
```

Comment:
Do we want `float`s close to integer values to be accepted?

Reply:
Good question. I always prefer to err on the side of:

- if it is common for the user to pass incorrect values (perhaps the frequencies have been generated programmatically, for example), then I would validate within epsilon precision

- If instead this is not a common workflow, I would prefer no validation, to avoid overhead

Reply:
I feel like programmatically obtaining the number of frequencies will usually be done by determining the `len` of some iterable or a variant of this. So I'd assume it is fair to skip validation.